File: config/shared_constants.js
/**
 * 共享常量加载器 - V1.0.0
 * ======================
 *
 * 从 config/shared_constants.json 加载全局共享常量，确保 Python/Node.js 一致性。
 *
 * V4.17: 建立跨语言常量统一标准
 *
 * 使用方法:
 *   const { MATCH_STATUS, MATCH_OUTCOME, HOME_WIN } = require('../config/shared_constants');
 *
 *   // 比赛状态
 *   const status = MATCH_STATUS.FINISHED;  // "finished"
 *
 *   // 比赛结果
 *   const result = MATCH_OUTCOME.HOME_WIN;  // 2
 *
 * @module config/shared_constants
 * @version V1.0.0
 * @since 2026-03-07
 */

'use strict';

const fs = require('fs');
const path = require('path');

// 加载共享常量 JSON
const CONSTANTS_PATH = path.join(__dirname, 'shared_constants.json');
let sharedConstants = null;

/**
 * 加载共享常量（带缓存）
 * @returns {Object} 共享常量对象
 */
function loadConstants() {
    if (sharedConstants) {
        return sharedConstants;
    }

    const content = fs.readFileSync(CONSTANTS_PATH, 'utf8');
    sharedConstants = JSON.parse(content);
    return sharedConstants;
}

/**
 * 重新加载常量（用于热更新）
 * @returns {Object} 共享常量对象
 */
function reloadConstants() {
    sharedConstants = null;
    return loadConstants();
}

// 加载常量
const constants = loadConstants();

// ============================================================================
// 比赛状态枚举
// ============================================================================

const MATCH_STATUS = Object.freeze({
    FINISHED: constants.match_status.FINISHED,      // "finished"
    LIVE: constants.match_status.LIVE,              // "live"
    SCHEDULED: constants.match_status.SCHEDULED,    // "scheduled"
    CANCELLED: constants.match_status.CANCELLED,    // "cancelled"
    COMPLETED: constants.match_status.COMPLETED,    // "completed"
    POSTPONED: constants.match_status.POSTPONED,    // "postponed"
    ABANDONED: constants.match_status.ABANDONED,    // "abandoned"

    /**
     * 检查状态是否为终态（不可变更）
     * @param {string} status - 状态值
     * @returns {boolean} 是否为终态
     */
    isTerminal(status) {
        const s = status?.toLowerCase();
        return s === this.FINISHED || s === this.CANCELLED || s === this.COMPLETED;
    },

    /**
     * 检查状态是否为活跃状态
     * @param {string} status - 状态值
     * @returns {boolean} 是否为活跃状态
     */
    isActive(status) {
        const s = status?.toLowerCase();
        return s === this.LIVE || s === this.SCHEDULED;
    }
});

// ============================================================================
// 比赛结果枚举
// ============================================================================

const MATCH_OUTCOME = Object.freeze({
    AWAY_WIN: constants.match_outcome.AWAY_WIN,  // 0
    DRAW: constants.match_outcome.DRAW,           // 1
    HOME_WIN: constants.match_outcome.HOME_WIN,   // 2

    /**
     * 根据比分确定结果
     * @param {number} homeScore - 主队得分
     * @param {number} awayScore - 客队得分
     * @returns {number} 结果 (0=客胜, 1=平局, 2=主胜)
     */
    fromScores(homeScore, awayScore) {
        if (homeScore > awayScore) return this.HOME_WIN;
        if (homeScore < awayScore) return this.AWAY_WIN;
        return this.DRAW;
    },

    /**
     * 获取所有标签（按顺序）
     * @returns {number[]} [0, 1, 2]
     */
    getAllLabels() {
        return [this.AWAY_WIN, this.DRAW, this.HOME_WIN];
    },

    /**
     * 获取标签名称
     * @param {number} label - 标签值
     * @returns {string} 标签名称
     */
    getLabelName(label) {
        const names = {
            [this.AWAY_WIN]: 'AWAY_WIN',
            [this.DRAW]: 'DRAW',
            [this.HOME_WIN]: 'HOME_WIN'
        };
        return names[label] || 'UNKNOWN';
    }
});

// ============================================================================
// 默认概率配置
// ============================================================================

const DEFAULT_PROBABILITIES = Object.freeze({
    HOME_WIN: constants.default_probabilities.HOME_WIN,  // 0.46
    DRAW: constants.default_probabilities.DRAW,           // 0.26
    AWAY_WIN: constants.default_probabilities.AWAY_WIN,   // 0.28

    /**
     * 转换为数组（按 AWAY_WIN, DRAW, HOME_WIN 顺序）
     * @returns {number[]} [0.28, 0.26, 0.46]
     */
    toArray() {
        return [this.AWAY_WIN, this.DRAW, this.HOME_WIN];
    },

    /**
     * 转换为对象
     * @returns {Object} 概率对象
     */
    toObject() {
        return {
            HOME_WIN: this.HOME_WIN,
            DRAW: this.DRAW,
            AWAY_WIN: this.AWAY_WIN
        };
    }
});

// ============================================================================
// 赔率限制配置
// ============================================================================

const ODDS_LIMITS = Object.freeze({
    MIN_ODDS: constants.odds_limits.min_odds,       // 1.01
    MAX_ODDS: constants.odds_limits.max_odds,       // 1000.0
    MIN_PAYOUT: constants.odds_limits.min_payout,   // 1.02
    MAX_PAYOUT: constants.odds_limits.max_payout,   // 1.08

    /**
     * 检查赔率是否在有效范围内
     * @param {number} odds - 赔率值
     * @returns {boolean} 是否有效
     */
    isValidOdds(odds) {
        return odds >= this.MIN_ODDS && odds <= this.MAX_ODDS;
    },

    /**
     * 检查返还率是否在有效范围内
     * @param {number} payout - 返还率
     * @returns {boolean} 是否有效
     */
    isValidPayout(payout) {
        return payout >= this.MIN_PAYOUT && payout <= this.MAX_PAYOUT;
    }
});

// ============================================================================
// 置信度阈值配置
// ============================================================================

const CONFIDENCE_THRESHOLDS = Object.freeze({
    HIGH: constants.confidence_thresholds.high,       // 0.70
    MEDIUM: constants.confidence_thresholds.medium,   // 0.55
    LOW: constants.confidence_thresholds.low,         // 0.40

    /**
     * 根据置信度获取等级
     * @param {number} confidence - 置信度
     * @returns {string} 等级名称
     */
    getLevel(confidence) {
        if (confidence >= this.HIGH) return 'HIGH';
        if (confidence >= this.MEDIUM) return 'MEDIUM';
        if (confidence >= this.LOW) return 'LOW';
        return 'VERY_LOW';
    }
});

// ============================================================================
// 向后兼容常量（直接导出数值）
// ============================================================================

// 比赛结果（整数）
const HOME_WIN = MATCH_OUTCOME.HOME_WIN;  // 2
const DRAW = MATCH_OUTCOME.DRAW;           // 1
const AWAY_WIN = MATCH_OUTCOME.AWAY_WIN;   // 0

// 比赛状态（字符串）
const STATUS_FINISHED = MATCH_STATUS.FINISHED;      // "finished"
const STATUS_LIVE = MATCH_STATUS.LIVE;              // "live"
const STATUS_SCHEDULED = MATCH_STATUS.SCHEDULED;    // "scheduled"
const STATUS_CANCELLED = MATCH_STATUS.CANCELLED;    // "cancelled"

// ============================================================================
// 导出
// ============================================================================

module.exports = {
    // 枚举对象
    MATCH_STATUS,
    MATCH_OUTCOME,

    // 配置对象
    DEFAULT_PROBABILITIES,
    ODDS_LIMITS,
    CONFIDENCE_THRESHOLDS,

    // 向后兼容常量（比赛结果）
    HOME_WIN,
    DRAW,
    AWAY_WIN,

    // 向后兼容常量（比赛状态）
    STATUS_FINISHED,
    STATUS_LIVE,
    STATUS_SCHEDULED,
    STATUS_CANCELLED,

    // 工具函数
    loadConstants,
    reloadConstants
};

File: src/constants/shared_constants.py
#!/usr/bin/env python3
"""
共享常量加载器 - V1.0.0
======================

从 config/shared_constants.json 加载全局共享常量，确保 Python/Node.js 一致性。

使用方法:
    from src.constants.shared_constants import (
        MATCH_STATUS, MATCH_OUTCOME, DEFAULT_PROBABILITIES
    )

    # 比赛状态
    status = MATCH_STATUS.FINISHED  # "finished"

    # 比赛结果
    result = MATCH_OUTCOME.HOME_WIN  # 2

Author: Systems Architect
Version: V1.0.0
Date: 2026-03-07
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

# 配置文件路径
SHARED_CONSTANTS_PATH = Path(__file__).parent.parent.parent / "config" / "shared_constants.json"


class MatchStatus(str, Enum):
    """比赛状态枚举"""

    FINISHED = "finished"
    LIVE = "live"
    SCHEDULED = "scheduled"
    CANCELLED = "cancelled"
    COMPLETED = "completed"
    POSTPONED = "postponed"
    ABANDONED = "abandoned"

    @classmethod
    def is_terminal(cls, status: str) -> bool:
        """检查状态是否为终态（不可变更）"""
        return status.lower() in (cls.FINISHED.value, cls.CANCELLED.value, cls.COMPLETED.value)

    @classmethod
    def is_active(cls, status: str) -> bool:
        """检查状态是否为活跃状态"""
        return status.lower() in (cls.LIVE.value, cls.SCHEDULED.value)


class MatchOutcome(int, Enum):
    """比赛结果枚举（整数，用于 ML 模型）"""

    AWAY_WIN = 0
    DRAW = 1
    HOME_WIN = 2

    @classmethod
    def from_scores(cls, home_score: int, away_score: int) -> MatchOutcome:
        """根据比分确定结果"""
        if home_score > away_score:
            return cls.HOME_WIN
        elif home_score < away_score:
            return cls.AWAY_WIN
        else:
            return cls.DRAW

    @classmethod
    def get_all_labels(cls) -> list[int]:
        """获取所有标签（按顺序）"""
        return [cls.AWAY_WIN, cls.DRAW, cls.HOME_WIN]

    @classmethod
    def get_label_names(cls) -> dict[int, str]:
        """获取标签名称映射"""
        return {cls.AWAY_WIN: "AWAY_WIN", cls.DRAW: "DRAW", cls.HOME_WIN: "HOME_WIN"}


@dataclass(frozen=True)
class DefaultProbabilities:
    """默认概率配置"""

    HOME_WIN: float = 0.46
    DRAW: float = 0.26
    AWAY_WIN: float = 0.28

    def to_dict(self) -> dict[str, float]:
        """转换为字典"""
        return {
            "HOME_WIN": self.HOME_WIN,
            "DRAW": self.DRAW,
            "AWAY_WIN": self.AWAY_WIN,
        }

    def to_list(self) -> list[float]:
        """转换为列表（按 AWAY_WIN, DRAW, HOME_WIN 顺序）"""
        return [self.AWAY_WIN, self.DRAW, self.HOME_WIN]


@dataclass(frozen=True)
class OddsLimits:
    """赔率限制配置"""

    MIN_ODDS: float = 1.01
    MAX_ODDS: float = 1000.0
    MIN_PAYOUT: float = 1.02
    MAX_PAYOUT: float = 1.08

    def is_valid_odds(self, odds: float) -> bool:
        """检查赔率是否在有效范围内"""
        return self.MIN_ODDS <= odds <= self.MAX_ODDS

    def is_valid_payout(self, payout: float) -> bool:
        """检查返还率是否在有效范围内"""
        return self.MIN_PAYOUT <= payout <= self.MAX_PAYOUT


@dataclass(frozen=True)
class ConfidenceThresholds:
    """置信度阈值配置"""

    HIGH: float = 0.70
    MEDIUM: float = 0.55
    LOW: float = 0.40

    def get_level(self, confidence: float) -> str:
        """根据置信度获取等级"""
        if confidence >= self.HIGH:
            return "HIGH"
        elif confidence >= self.MEDIUM:
            return "MEDIUM"
        elif confidence >= self.LOW:
            return "LOW"
        else:
            return "VERY_LOW"


@dataclass(frozen=True)
class StatisticalThresholds:
    """统计阈值配置"""

    MIN_SAMPLE_SIZE: int = 10
    RELIABLE_SAMPLE_SIZE: int = 30
    LARGE_SAMPLE_SIZE: int = 100

    def get_reliability(self, sample_size: int) -> str:
        """根据样本量获取可靠性等级"""
        if sample_size >= self.LARGE_SAMPLE_SIZE:
            return "HIGH"
        elif sample_size >= self.RELIABLE_SAMPLE_SIZE:
            return "MEDIUM"
        elif sample_size >= self.MIN_SAMPLE_SIZE:
            return "LOW"
        else:
            return "INSUFFICIENT"


def _load_shared_constants() -> dict[str, Any]:
    """加载共享常量 JSON 文件"""
    if not SHARED_CONSTANTS_PATH.exists():
        raise FileNotFoundError(f"Shared constants file not found: {SHARED_CONSTANTS_PATH}")

    with open(SHARED_CONSTANTS_PATH, encoding="utf-8") as f:
        return json.load(f)


# 全局实例（延迟加载）
_constants_cache: dict[str, Any] | None = None


def get_constants() -> dict[str, Any]:
    """获取共享常量（带缓存）"""
    global _constants_cache
    if _constants_cache is None:
        _constants_cache = _load_shared_constants()
    return _constants_cache


def reload_constants() -> dict[str, Any]:
    """重新加载常量（用于热更新）"""
    global _constants_cache
    _constants_cache = _load_shared_constants()
    return _constants_cache


# 导出的常量实例
# V4.46.2: Enum 类直接使用，dataclass 类需要实例化
MATCH_STATUS = MatchStatus  # Enum 类，直接使用
MATCH_OUTCOME = MatchOutcome  # Enum 类，直接使用
DEFAULT_PROBABILITIES = DefaultProbabilities()  # dataclass，实例化
ODDS_LIMITS = OddsLimits()  # dataclass，实例化
CONFIDENCE_THRESHOLDS = ConfidenceThresholds()  # dataclass，实例化
STATISTICAL_THRESHOLDS = StatisticalThresholds()  # dataclass，实例化

# 向后兼容的常量值
HOME_WIN = MatchOutcome.HOME_WIN  # 2
DRAW = MatchOutcome.DRAW  # 1
AWAY_WIN = MatchOutcome.AWAY_WIN  # 0

# 比赛状态常量（字符串）
STATUS_FINISHED = MatchStatus.FINISHED.value  # "finished"
STATUS_LIVE = MatchStatus.LIVE.value  # "live"
STATUS_SCHEDULED = MatchStatus.SCHEDULED.value  # "scheduled"
STATUS_CANCELLED = MatchStatus.CANCELLED.value  # "cancelled"

__all__ = [
    # 枚举类
    "MatchStatus",
    "MatchOutcome",
    # 配置类
    "DefaultProbabilities",
    "OddsLimits",
    "ConfidenceThresholds",
    "StatisticalThresholds",
    # 全局实例
    "MATCH_STATUS",
    "MATCH_OUTCOME",
    "DEFAULT_PROBABILITIES",
    "ODDS_LIMITS",
    "CONFIDENCE_THRESHOLDS",
    "STATISTICAL_THRESHOLDS",
    # 向后兼容常量
    "HOME_WIN",
    "DRAW",
    "AWAY_WIN",
    "STATUS_FINISHED",
    "STATUS_LIVE",
    "STATUS_SCHEDULED",
    "STATUS_CANCELLED",
    # 工具函数
    "get_constants",
    "reload_constants",
]
